src/Cardano/Wallet/Primitive/Signing.hs

```diff
@@ -0,0 +1,135 @@
+module Cardano.Wallet.Primitive.Signing
+    ( -- * Sign transactions
+      mkStdTx
+
+      -- * Helpers
+    , shuffle
+    )
+    where
+
+
+import Prelude
+
+import Cardano.Wallet.Binary
+    ( TxWitness (..), encodeTx, toByteString )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( ChangeChain (ExternalChain)
+    , Depth (AddressK, RootK)
+    , Key
+    , Passphrase
+    , SignTag (SignTx)
+    , XPrv
+    , deriveAccountPrivateKey
+    , deriveAddressPrivateKey
+    , encodeXPub
+    , publicKey
+    , sign
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( SeqState (externalPool), lookupAddress )
+import Cardano.Wallet.Primitive.Types
+    ( Address, Hash (..), Tx (..), TxIn, TxOut (..) )
+import Control.Monad
+    ( forM, forM_ )
+import Crypto.Hash
+    ( Blake2b_256, hash )
+import Crypto.Number.Generate
+    ( generateBetween )
+import Data.ByteArray
+    ( convert )
+import Data.ByteString
+    ( ByteString )
+import Data.Vector.Mutable
+    ( IOVector )
+
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as MV
+
+-- | Build a transaction
+
+-- | Construct a standard transaction
+--
+-- " Standard " here refers to the fact that we do not deal with redemption,
+-- multisignature transactions, etc.
+--
+-- TODO: I removed FakeSigner/SafeSigner. Might be wrong.
+mkStdTx :: SeqState
+        -> (Key 'RootK XPrv, Passphrase "encryption")
+--        -> ([TxOut] -> m [TxOut])
+        -> [(TxIn, TxOut)]
+        -- ^ Selected inputs
+        -> [TxOut]
+        -- ^ Selected outputs (including change)
+        -> Maybe (Tx, [TxWitness])
+mkStdTx seqState (rootPrv, pwd) ownedIns outs = do
+
+    let ins = (fmap fst ownedIns)
+        tx = Tx ins outs
+        txSigData = hashTx tx
+
+    txWitnesses <- forM ownedIns (\(_in, TxOut addr _c) ->
+        mkWitness txSigData <$> keyFrom addr)
+
+    return (tx, txWitnesses)
+
+  where
+    hashTx :: Tx -> Hash "tx"
+    hashTx tx = Hash
+        $ convert
+        $ (hash @ByteString @Blake2b_256)
+        $ toByteString
+        $ encodeTx tx
+
+    keyFrom :: Address -> (Maybe (Key 'AddressK XPrv))
+    keyFrom addr = do
+        -- We are assuming there is only one account
+        let account = minBound
+        let accountPrv = deriveAccountPrivateKey pwd rootPrv account
+
+        -- We are ignoring the new state/pool. We won't discover any new
+        -- addresses when submitting transactions.
+        index <- fst $ lookupAddress addr (externalPool seqState)
+
+        return $ deriveAddressPrivateKey pwd accountPrv ExternalChain index
+
+    mkWitness :: Hash "tx" -> Key 'AddressK XPrv -> TxWitness
+    mkWitness tx xPrv =
+        PublicKeyWitness $
+            encodeXPub (publicKey xPrv) <>
+            getHash (sign (SignTx tx) (xPrv, pwd))
+
+
+-- | Shuffles a list. Meant to be passed to @mkStdTx@.
+--
+-- Implementation was copied from cardano-sl
+shuffle :: [a] -> IO [a]
+shuffle = modifyInPlace $ \v -> do
```

Comment:
I think we want to move that in the coin selection :thinking:

Reply:
Hm, ok